---
title: "CITE-SeqPreprocessing"
output: html_document
---

```{r}
#Load library
library(Seurat)
library(dplyr)
library(ggplot2)
library(RColorBrewer)

#Analysis for all patient
all <- readRDS("rawdata/Complete_integration_obj.rds")

```

```{r}

#Filter negative and doublet from analysis
all_filter <- subset(all,cells =row.names(all@meta.data[!grepl("Doublet|Negative", all@meta.data$sample_ID),]))#subset(all, subset = SNP_sample_ID != "Doublet" | SNP_sample_ID != "Negative")
print( unique(all_filter@meta.data$sample_ID))

#Sample ID should contains the names of all samples if mouse/genative/doublet are removed.
#> unique(all_filter@meta.data$sample_ID)
#[1] "MELCAP01_MIA01_LN_Pre"        "MELCAP01_MIA02_Subcut_TBC"   
#[3] "MELCAP02_MIA11_Subcut_Pre-1"  "MELCAP02_MIA01_LN_Prog"      
#[5] "MELCAP03_MIA11-Subcut_EDT-3"  "MELCAP03_MIA09_LN_Pre"       
#[7] "MELCAP04_MIA10_Subcut_Pre"    "MELCAP04_MIA11-Subcut_Prog-2"
```

```{r setup, warning=FALSE}
DefaultAssay(all_filter) <- "RNA"
all_filter <- all_filter[!grepl("mm10", row.names(all_filter)),]
all_filter <- all_filter[!(grepl("RP", row.names(all_filter)) & grepl("-", row.names(all_filter))),]

immx.list <- SplitObject(all_filter, split.by = "sample_ID")
immx.list <- lapply(X = immx.list, FUN = SCTransform)
features <- SelectIntegrationFeatures(object.list = immx.list, nfeatures = 5000)
immx.list <- PrepSCTIntegration(object.list = immx.list, anchor.features = features)
```


```{r}
immune.anchors <- FindIntegrationAnchors(object.list = immx.list, normalization.method = "SCT",
    anchor.features = features)
immune.combined.sct <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT")
immune.combined.sct <- RunPCA(immune.combined.sct, verbose = FALSE)
immune.combined.sct <- RunUMAP(immune.combined.sct, reduction = "pca", dims = 1:30)
p1 <- DimPlot(immune.combined.sct, reduction = "umap", group.by = "sample_ID")

```
```{r}
immune.combined.sct <- FindNeighbors(immune.combined.sct)

immune.combined.sct <- FindClusters(immune.combined.sct, algorithm = 1, resolution = 1)

```
```{r}
DimPlot(immune.combined.sct, reduction = "umap", group.by = "sample_ID")
DimPlot(immune.combined.sct, reduction = "umap", group.by = "integrated_snn_res.1")
```


```{r}
#hpca.se <- celldex::HumanPrimaryCellAtlasData()
#hpca.se
library(SingleR)

immune.ref <- MonacoImmuneData()
#immunestroma.ref <- BlueprintEncodeData()
iCE <- DatabaseImmuneCellExpressionData()
#convert seurat object to singleR obj
objR <- as.SingleCellExperiment(immune.combined.sct)

```
```{r}
#Run the databases
#pred.hpca <- SingleR(test = objR, ref = hpca.se, assay.type.test=1, labels = hpca.se$label.main)
pred.immune.ref <- SingleR(test = objR, ref = immune.ref, assay.type.test=1, labels = immune.ref$label.fine, clusters=immune.combined.sct@meta.data$integrated_snn_res.1)
#pred.immunestroma.ref <- SingleR(test = objR, ref = immunestroma.ref, assay.type.test=1, labels =immunestroma.ref$label.fine)
```

```{r}
iCE <- DatabaseImmuneCellExpressionData()
pred.iCE <- SingleR(test = objR, ref = iCE, assay.type.test=1, labels =iCE$label.fine, clusters=immune.combined.sct@meta.data$integrated_snn_res.1)

```


```{r}
png("UMAP-iRNA.png",width=12, height=16,res=300, units = 'in')

immune.combined.sct@meta.data[,'SingleR_ann'] <- pred.iCE[immune.combined.sct@meta.data$integrated_snn_res.1,'pruned.labels']#pred.immune.ref$pruned.labels#pred.iCE$pruned.labels

immune.combined.sct@meta.data$immune_refNA <- is.na(immune.combined.sct@meta.data$SingleR_ann)
immune.combined.sct1 <- subset(x = immune.combined.sct, subset = (immune_refNA == FALSE ))

 #DimPlot(immune.combined.sct, reduction = "umap", group.by = "sample_ID")

 DimPlot(immune.combined.sct1, reduction = "umap", group.by = "SingleR_ann", label=T, repel=TRUE, split.by = "sample_ID", pt.size=1,ncol=2) + NoLegend()
#p1 + p2
dev.off()
```


```{r}
adt_filter <- subset(all,cells =row.names(all@meta.data[!grepl("Doublet|Negative", all@meta.data$sample_ID),]))#subset(all, subset = SNP_sample_ID != "Doublet" | SNP_sample_ID != "Negative")
print( unique(adt_filter@meta.data$sample_ID))
DefaultAssay(adt_filter) <- "ADT"
adt_filter <- adt_filter[!grepl("Rat|Mouse|Arm.ham", row.names(adt_filter)),]


```
```{r}
mapDF <- read.csv('ADTProteins.txt',sep=',')
rownames(mapDF) <- mapDF$marker
mapDF
```
```{r}
presentNames <- rownames(adt_filter)[(rownames(adt_filter) %in% mapDF$marker)]
adt_filter <- adt_filter[presentNames,]

mapDF[presentNames,'gene_symbol']
```
```{r}
immx.list <- SplitObject(adt_filter, split.by = "sample_ID")
immx.list <- lapply(X = immx.list, FUN = SCTransform, assay="ADT")
features <- SelectIntegrationFeatures(object.list = immx.list, nfeatures = 5000)
immx.list <- PrepSCTIntegration(object.list = immx.list, anchor.features = features)
immune.anchors <- FindIntegrationAnchors(object.list = immx.list, normalization.method = "SCT",
    anchor.features = features)
immune.combined.ADTsct <- IntegrateData(anchorset = immune.anchors, normalization.method = "SCT")
immune.combined.ADTsct <- RunPCA(immune.combined.ADTsct, verbose = FALSE)
immune.combined.ADTsct <- RunUMAP(immune.combined.ADTsct, reduction = "pca", dims = 1:30)
p1 <- DimPlot(immune.combined.ADTsct, reduction = "umap", group.by = "sample_ID")

immune.combined.ADTsct <- FindNeighbors(immune.combined.ADTsct)

immune.combined.ADTsct <- FindClusters(immune.combined.ADTsct, algorithm = 1, resolution = 1)

```
```{r}
DimPlot(immune.combined.ADTsct, reduction = "umap", group.by = "sample_ID")
DimPlot(immune.combined.ADTsct, reduction = "umap", group.by = "integrated_snn_res.1")
```

```{r}
objRADT <- as.SingleCellExperiment(immune.combined.ADTsct)
row.names(objRADT) <- mapDF[rownames(objRADT),'gene_symbol']
```


```{r}
#ADT.pred.iref <- SingleR(test = objRADT, ref = immune.ref, assay.type.test=1, labels = immune.ref$label.fine,clusters=immune.combined.ADTsct@meta.data$integrated_snn_res.1)
ADT.pred.iref <- SingleR(test = objRADT, ref = iCE, assay.type.test=1, labels = iCE$label.fine,clusters=immune.combined.ADTsct@meta.data$integrated_snn_res.1)

png("UMAP-iADT.png",width=12, height=16,res=300, units = 'in')

immune.combined.ADTsct@meta.data[,'SingleR_ann'] <- ADT.pred.iref[immune.combined.ADTsct@meta.data$integrated_snn_res.1,'pruned.labels']#ADT.pred.iref$pruned.labels#pred.iCE$pruned.labels

immune.combined.ADTsct@meta.data$immune_refNA <- is.na(immune.combined.ADTsct@meta.data$SingleR_ann)
immune.combined.ADTsct1 <- subset(x = immune.combined.ADTsct, subset = (immune_refNA == FALSE ))

 #DimPlot(immune.combined.sct, reduction = "umap", group.by = "sample_ID")

 DimPlot(immune.combined.ADTsct1, reduction = "umap", group.by = "SingleR_ann", label=T, repel=TRUE, split.by = "sample_ID", pt.size=1,ncol=2) + NoLegend()
#p1 + p2
dev.off()
```


```{r}
#/////////////////redo anchoring on IMMUNE cells

# Check names - since measurements were made in the same cells, the two matrices have identical column names
all.equal(colnames(objR), colnames(objRADT))

#metadata
all_meta <- as.data.frame(all@meta.data)
idx <- match(colnames(objR), rownames(all_meta))
reordered_mdata <- all_meta[idx,]
all(row.names(reordered_mdata) == colnames(objR))

# creates a Seurat object based on the scRNA-seq data
ptall <- CreateSeuratObject(counts = objR@assays@data$logcounts, meta.data = reordered_mdata)
adt_assay <- CreateAssayObject(counts = objRADT@assays@data$logcounts)
ptall[["ADT"]] <- adt_assay

# Validate that the object now contains multiple assays
Assays(ptall)
#> Assays(ptall)
#[1] "RNA" "ADT"
```

```{r}
#Running WNN
DefaultAssay(ptall) <- 'RNA'
#ptall <- NormalizeData(ptall) %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA()
VariableFeatures(ptall) <- rownames(ptall[["RNA"]])
ptall <- SetAssayData(ptall, assay = "RNA", slot = "scale.data", new.data = as.matrix(objR@assays@data$logcounts))

ptall <-  RunPCA(ptall,reduction.name = "rnapca")

DefaultAssay(ptall) <- 'ADT'
ptall <- SetAssayData(ptall, assay = "ADT", slot = "scale.data", new.data = as.matrix(objRADT@assays@data$logcounts))

VariableFeatures(ptall) <- rownames(ptall[["ADT"]])

#ptall <- NormalizeData(ptall, normalization.method = "CLR", margin = 2) %>% 
#  ScaleData() %>% RunPCA(reduction.name = "apca")
ptall <- RunPCA(ptall,reduction.name = "adtpca")

ptall <- FindMultiModalNeighbors(
  ptall, reduction.list = list("rnapca", "adtpca"), 
  dims.list = list(1:25, 1:18), modality.weight.name = "RNA.weight")
  
ptall <- RunUMAP(ptall, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
ptall <- FindClusters(ptall, graph.name = "wsnn", algorithm = 1, resolution = 1)
ptall <- RunUMAP(ptall, reduction = 'rnapca', dims = 1:30, assay = 'RNA', 
              reduction.name = 'rna.umap', reduction.key = 'rnaUMAP_')
ptall <- RunUMAP(ptall, reduction = 'adtpca', dims = 1:18, assay = 'ADT', 
              reduction.name = 'adt.umap', reduction.key = 'adtUMAP_')


```
```{r}
pdf("UMAP-iWNN_all_allcells.pdf")
DimPlot(ptall, reduction = "wnn.umap", label = FALSE)
DimPlot(ptall, reduction = "wnn.umap", label = TRUE)
DimPlot(ptall, reduction = "wnn.umap", group.by = "sample_ID", label = FALSE)
DimPlot(ptall, reduction = "wnn.umap", group.by = "sample_ID", label = TRUE)
DimPlot(ptall, reduction = "wnn.umap", split.by = "sample_ID", label = FALSE)
DimPlot(ptall, reduction = "wnn.umap", split.by = "sample_ID", label = TRUE)
dev.off()
p3 <- DimPlot(ptall, reduction = 'rna.umap', label = TRUE, 
              repel = TRUE, label.size = 2.5) + ggtitle("RNA_scRNAseq")
p4 <- DimPlot(ptall, reduction = 'adt.umap', label = TRUE, 
              repel = TRUE, label.size = 2.5) + ggtitle("Protein_CITEseq")
p5 <- DimPlot(ptall, reduction = "wnn.umap",
              label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle("WNN")
png("UMAP-all_allcells25-15-0.8.png", width=35, height=15, units="cm", res=500)
p3 + p4 + p5
dev.off()


DefaultAssay(ptall) <- "ADT"
p6 <- FeaturePlot(ptall, features = c("CITE-CD3", "CITE-CD8A", "CITE-PD1", "CITE-CTLA4"),
                  reduction = 'wnn.umap', max.cutoff = 2, 
                  cols = c("lightgrey","darkgreen"), ncol = 4)
DefaultAssay(ptall) <- 'RNA'
p7 <- FeaturePlot(ptall, features = c("CD3E", "CD8A", "PDCD1", "CTLA4"), 
                  reduction = 'wnn.umap', max.cutoff = 2, ncol = 4)
png("Featureplot_allcells_iadt.png", width=40, height=15, units="cm", res=500)
p6
dev.off()
png("Featureplot_allcells_iRNA.png", width=40, height=15, units="cm", res=500)
p7
dev.off()

#Find all markers
#ptall.adtmarkers <- FindAllMarkers(ptall, min.pct = 0.15, logfc.threshold = 0.15, assay = "ADT")
#ptall.rnamarkers <- FindAllMarkers(ptall, min.pct = 0.15, logfc.threshold = 0.15, assay = "RNA")

#Find top 20 markers
#adt
#top20adt <- ptall.adtmarkers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
#rna
#top20rna <- ptall.rnamarkers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
```

```{r}
saveRDS(ptall,'CITE-Seq_wnn_integrated.RDS')
```

```{r}
#Find top 20 markers
#adt
top20adt <- ptall.adtmarkers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
#rna
top20rna <- ptall.rnamarkers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
```


```{r}
library(AnnotationHub)
hsdb = AnnotationHub()[['AH73881']]

hs.exons <- exonsBy(hsdb, by="gene")
hs.exons <- reduce(hs.exons)
hs.len <- sum(width(hs.exons))
```

## Including Plots
```{r}
library('biomaRt')
#mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- names(hs.len)
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id",
"hgnc_symbol"),values=genes,mart= mart)
```



```{r}
head(G_list)
```
```{r}
uniqGList <- G_list[!duplicated(G_list$hgnc_symbol),]
keep <- row.names(all_filter@assays$RNA)[row.names(all_filter@assays$RNA) %in% uniqGList$hgnc_symbol]
rownames(uniqGList) <- uniqGList$hgnc_symbol
DefaultAssay(immune.combined.sct) <- "RNA"
```
```{r}
library(SingleR)
library(scater)
tpmMIA <- calculateTPM(immune.combined.sct@assays$RNA[keep,], lengths=hs.len[uniqGList[keep,'ensembl_gene_id']])

```

```{r}
iCE <-BlueprintEncodeData()#DatabaseImmuneCellExpressionData()

# Performing the assignment.
iRNA.pred.iref <- SingleR(test = tpmMIA, ref = iCE, assay.type.test=1, labels = iCE$label.fine,
                          clusters=ptall@meta.data$wsnn_res.1)

png("UMAP-RNARaw.png",width=12, height=16,res=300, units = 'in')

ptall@meta.data[,'SingleR_RNARaw_ann'] <- iRNA.pred.iref[ptall@meta.data$wsnn_res.1,'pruned.labels']

ptall@meta.data$immune_refNA <- is.na(ptall@meta.data$SingleR_RNARaw_ann)
ptall2 <- subset(x = ptall, subset = (immune_refNA == FALSE ))

 #DimPlot(immune.combined.sct, reduction = "umap", group.by = "sample_ID")

 DimPlot(ptall2, reduction = "wnn.umap", group.by = "SingleR_RNARaw_ann", label=T, repel=TRUE, split.by = "sample_ID", pt.size=1,ncol=2) + NoLegend()
#p1 + p2
dev.off()
```
```{r}
saveRDS(ptall,'CITE-Seq_wnn_integrated_annotated.RDS')

```

```{r}
write.csv(ptall@meta.data,'MIA_annotatedMetadata.csv',)
```
```{r}
write.csv(tpmMIA,'MIA_RNA_TPM.csv')
```

```{r}
write.csv(adt_filter@assays$ADT@counts,'MIA_ADT_Raw.csv')
write.csv(ptall@assays$ADT@counts,'MIA_ADT_SCT.csv')
write.csv(ptall@assays$RNA@counts,'MIA_RNA_SCT.csv')
```

